orchagent/fabricportsorch.cpp

```diff
@@ -0,0 +1,338 @@
+
+                queue_attrs[0].id = SAI_QUEUE_ATTR_TYPE;
+                queue_attrs[1].id = SAI_QUEUE_ATTR_INDEX;
+                status = sai_queue_api->get_queue_attribute(m_queue_ids[q], 2, queue_attrs);
+                if (status == SAI_STATUS_SUCCESS)
+                {
+                    string queueType;
+                    string queueIndex;
+                    switch (queue_attrs[0].value.s32)
+                    {
+                        case SAI_QUEUE_TYPE_ALL:
+                            queueType = "SAI_QUEUE_TYPE_ALL";
+                            break;
+                        case SAI_QUEUE_TYPE_UNICAST:
+                            queueType = "SAI_QUEUE_TYPE_UNICAST";
+                            break;
+                        case SAI_QUEUE_TYPE_MULTICAST:
+                            queueType = "SAI_QUEUE_TYPE_MULTICAST";
+                            break;
+                        default:
+                            throw runtime_error("Got unsupported queue type");
+                    }
+                    queueTypeMap.emplace_back(queue, queueType);
+                    queueIndex = to_string(queue_attrs[1].value.u8);
+                    queueIndexMap.emplace_back(queue, queueIndex);
+                }
+                else
+                {
+                    SWSS_LOG_NOTICE("FabricPortsOrch cannot get fabric port queue type and index");
+                }
+
+                std::unordered_set<string> counter_stats;
+                for (const auto& it: queue_stat_ids)
+                {
+                    counter_stats.emplace(sai_serialize_queue_stat(it));
+                }
+                queue_stat_manager.setCounterIdList(m_queue_ids[q], CounterType::QUEUE, counter_stats);
+            }
+
+            m_portNameQueueCounterTable->set("", portNameQueueMap);
+            m_queuePortCounterTable->set("", queuePortMap);
+            m_queueTypeCounterTable->set("", queueTypeMap);
+            m_queueIndexCounterTable->set("", queueIndexMap);
+        }
+    }
+
+    m_isQueueStatsGenerated = true;
+}
+
+void FabricPortsOrch::updateFabricPortState()
+{
+    if (!m_getFabricPortListDone) return;
+
+    sai_status_t status;
+    sai_attribute_t attr;
+
+    for (auto p : m_fabricLanePortMap)
+    {
+        int lane = p.first;
+        sai_object_id_t port = p.second;
+
+        string key = "PORT" + to_string(lane);
+        std::vector<FieldValueTuple> values;
+
+        attr.id = SAI_PORT_ATTR_FABRIC_ATTACHED;
+        status = sai_port_api->get_port_attribute(port, 1, &attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            throw runtime_error("FabricPortsOrch get port status failure");
+        }
+        FieldValueTuple s("STATUS", attr.value.booldata ? "up" : "down");
+        values.push_back(s);
+
+        if (attr.value.booldata)
+        {
+            attr.id = SAI_PORT_ATTR_FABRIC_ATTACHED_SWITCH_ID;
```

Comment:
Add the crc error collection here.

Reply:
Can we detect if a link flapped between the polling period? Is that a count that can be maintained in STATE_DB?

Reply:
I will add counter list, one counter for one port, that counts the number of times the link flaps. I also maintain the last time the link flaps.

Reply:
CRC error collection will be added in another PR. It's because sonic-sairedis now doesn't understand attribute type which is used for getting fabric port error list. So we will need to modify sonic-sairedis, and then add CRC in sonic-swss.

Reply:
Ok. Sounds good to me.